crypto/bn/rsaz_exp_x2.c

```diff
@@ -229,149 +264,207 @@ int RSAZ_mod_exp_avx512_x2(BN_ULONG *res1,
      */
-    red_X[0][0] = 1;
-    red_X[1][0] = 1;
-    DAMM(red_table[0][0], (const BN_ULONG*)red_X, rr, m, k0);
-    DAMM(red_table[1][0], base,  rr, m, k0);
-
-    for (idx = 1; idx < (int)((1U << EXP_WIN_SIZE) / 2); idx++) {
-        DAMS(red_table[2 * idx + 0][0], red_table[1 * idx][0], m, k0);
-        DAMM(red_table[2 * idx + 1][0], red_table[2 * idx][0], red_table[1][0], m, k0);
+    red_X[0 * red_digits] = 1;
+    red_X[1 * red_digits] = 1;
+    damm(&red_table[0 * 2 * red_digits], (const BN_ULONG*)red_X, rr, m, k0);
+    damm(&red_table[1 * 2 * red_digits], base,  rr, m, k0);
+
+    for (idx = 1; idx < (int)((1U << exp_win_size) / 2); idx++) {
+        DAMS(&red_table[(2 * idx + 0) * 2 * red_digits],
+             &red_table[(1 * idx)     * 2 * red_digits], m, k0);
+        damm(&red_table[(2 * idx + 1) * 2 * red_digits],
+             &red_table[(2 * idx)     * 2 * red_digits],
+             &red_table[1 * 2 * red_digits], m, k0);
     }
 
     /* Copy and expand exponents */
-    memcpy(expz[0], exp[0], EXP_DIGITS * sizeof(BN_ULONG));
-    expz[0][EXP_DIGITS] = 0;
-    memcpy(expz[1], exp[1], EXP_DIGITS * sizeof(BN_ULONG));
-    expz[1][EXP_DIGITS] = 0;
+    memcpy(&expz[0 * (exp_digits + 1)], exp[0], exp_digits * sizeof(BN_ULONG));
+    expz[1 * (exp_digits + 1) - 1] = 0;
+    memcpy(&expz[1 * (exp_digits + 1)], exp[1], exp_digits * sizeof(BN_ULONG));
+    expz[2 * (exp_digits + 1) - 1] = 0;
 
     /* Exponentiation */
     {
-        int rem = BITSIZE_MODULUS % EXP_WIN_SIZE;
-        int delta = rem ? rem : EXP_WIN_SIZE;
-        BN_ULONG table_idx_mask = EXP_WIN_MASK;
+        int rem = modulus_bitsize % exp_win_size;
+        int delta = rem ? rem : exp_win_size;
+        BN_ULONG table_idx_mask = exp_win_mask;
 
-        int exp_bit_no = BITSIZE_MODULUS - delta;
+        int exp_bit_no = modulus_bitsize - delta;
         int exp_chunk_no = exp_bit_no / 64;
         int exp_chunk_shift = exp_bit_no % 64;
 
         /* Process 1-st exp window - just init result */
-        BN_ULONG red_table_idx_0 = expz[0][exp_chunk_no];
-        BN_ULONG red_table_idx_1 = expz[1][exp_chunk_no];
+        BN_ULONG red_table_idx_0 = expz[exp_chunk_no + 0 * (exp_digits + 1)];
+        BN_ULONG red_table_idx_1 = expz[exp_chunk_no + 1 * (exp_digits + 1)];
         /*
          * The function operates with fixed moduli sizes divisible by 64,
          * thus table index here is always in supported range [0, EXP_WIN_SIZE).
          */
         red_table_idx_0 >>= exp_chunk_shift;
         red_table_idx_1 >>= exp_chunk_shift;
 
-        extract_multiplier_2x20_win5(red_Y[0], (const BN_ULONG*)red_table, (int)red_table_idx_0, 0);
-        extract_multiplier_2x20_win5(red_Y[1], (const BN_ULONG*)red_table, (int)red_table_idx_1, 1);
+        extract(&red_Y[0 * red_digits], (const BN_ULONG*)red_table, (int)red_table_idx_0, 0);
+        extract(&red_Y[1 * red_digits], (const BN_ULONG*)red_table, (int)red_table_idx_1, 1);
```

Comment:
There are 2 extract() function calls back to back here, which walk through the exact same table of data (red_table), checking every second precomputation, and offset by 1 precomputation (tbl_idx) for each call.

A dual_extract() routine may be more efficient, which would walk through all precomputes sequentially once, and checking for the desired precompute for tbl_idx 0 and 1 alternately.

There could be nicer caching behaviour, since precomputes are not always multiples of cache line size which mightn't be friendly for reading every second precompute.

extract() function calls are separated in the logic just below (450 and 467), but the logic could easily be modified such that a dual_extract could be used

Reply:
reviewed the updated code for 2K, 3K and 4K which all looks good to me